### PR TITLE
Range requests

### LINKED ISSUE #734: Support `Range` requests on `/assets/bytes/{path}`
## Use case

```py
from tiled.client import from_profile
import pandas

c = from_profile('local', api_key='secret')
df = pandas.DataFrame({"A": [1,2,3], "B": [4,5,6], "C": [7,8,9]})
c.write_dataframe(df, key="y")

import duckdb
db = duckdb.connect(':memory:')
db.execute('install httpfs; load httpfs;')
db.execute('SELECT * FROM read_parquet("http://localhost:8000/api/v1/asset/bytes/y?id=1&api_key=secret")')  # uses Range header
```

## Background

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Range
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Range
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Range
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/206
https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests

Starlette has postponed consideration of support for this. It might make sense to vendor a copy of this approach for now: https://github.com/encode/starlette/pull/1999

### PR DESCRIPTION
**Pull request by danielballan**
As noted in #734 the motivating use cases here was DuckDB's httpfs, which tries to do partial access of files over HTTP. Given a server with one DataFrame in it:

```sh
tiled serve catalog --temp
```

```py
from tiled.cilent import from_uri
c = from_uri('http://localhost:8000', api_key='secret')
df = pandas.DataFrame({"A": [1,2,3], "B": [4,5,6], "C": [7,8,9]})
c.write_dataframe(df, key="x")
```

DuckDB's httpfs plugin issues requests that use the `Range` header:

```py
import duckdb
db = duckdb.connect(':memory:')
db.execute('install httpfs; load httpfs;')
db.execute('SELECT * FROM read_parquet("http://localhost:8000/api/v1/asset/bytes/x?id=1&api_key=secret")').fetchall()
```

generating these server logs confirming that Partial Access was invoked:

```
[531d1958246ae636] 127.0.0.1:59854 (public) - "HEAD /api/v1/asset/bytes/x?id=1&api_key=secret HTTP/1.1" 405 Method Not Allowed
[1e2951b2eaa982dd] 127.0.0.1:59854 (admin) - "GET /api/v1/asset/bytes/x?id=1&api_key=secret HTTP/1.1" 206 Partial Content
[6a5b9f06acf0a7e5] 127.0.0.1:59854 (admin) - "GET /api/v1/asset/bytes/x?id=1&api_key=secret HTTP/1.1" 206 Partial Content
[79d13fc32a84235e] 127.0.0.1:59854 (admin) - "GET /api/v1/asset/bytes/x?id=1&api_key=secret HTTP/1.1" 206 Partial Content
[90ebff815c556fa1] 127.0.0.1:59866 (admin) - "GET /api/v1/asset/bytes/x?id=1&api_key=secret HTTP/1.1" 206 Partial Content
```

### Checklist
- [x] Add a Changelog entry
- [x] Add the ticket number which this PR closes to the comment section

Closes #734

### CODE REVIEW

File: tiled/server/file_response_with_range.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# This is adapted from a closed PR in starlette which was reviewed by
+# a core maintainer but put aside for now in favor of other priorities.
```
Comment (hyperrealist) [inline tiled/server/file_response_with_range.py:2]: Perhaps not put aside anymore?

File: tiled/server/file_response_with_range.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        if self.stat_result is None:
+            try:
+                stat_result = await anyio.to_thread.run_sync(os.stat, self.path)
+                self.set_stat_headers(stat_result)
+            except FileNotFoundError:
+                raise RuntimeError(f"File at path {self.path} does not exist.")
+            else:
+                mode = stat_result.st_mode
+                if not stat.S_ISREG(mode):
+                    raise RuntimeError(f"File at path {self.path} is not a file.")
```
Comment (hyperrealist) [inline tiled/server/file_response_with_range.py:63]: I wouldn't raise this error message verbatim with a straight face 😄

**Review by hyperrealist — APPROVED** [review]
Looks good to me, with the caveat that I could only do a shallow review of `file_response_with_range.py` given my inexperience with starlette. But the logic checks out.

I edited `file_response_with_range.py` to replace literal http status codes with `HTTP_200__OK` et. al. from starlette.status.

Also there are few minor comments in the files.

File: tiled/server/file_response_with_range.py (hunk 1 above)
Comment (danielballan) [reply tiled/server/file_response_with_range.py]: Ah, I meant put aside by starlette's core devs, not by us. I revised the comment to clarify and add a little more  context.

**Pull request merged**